geoalchemy2/comparator.py

```diff
@@ -86,7 +89,15 @@ def __getattr__(self, name):
         # This is not to mess up with SQLAlchemy's use of
         # hasattr/getattr on Column objects.
 
-        if not name.startswith('ST_'):
+        if (
+            (
+                SA_VERSION < [1, 3, 4]
+                and not name.startswith('ST_')
+            ) or (
+                SA_VERSION >= [1, 3, 4]
+                and not name.lower().startswith('st_')
+            )
+        ):
```

Comment:
Why not using `if not name.lower().startswith('st_')` with all versions of SQLAlchemy? What problem could that cause?

I tend to think that it's ok to use `if not name.lower().startswith('st_')` with all versions of SQLAlchemy. With SQLAlchemy 1.3.3 and lower it would mean that we accept both ST_ and st_ functions. I think it's ok, even though the st_ (lower-case) version is not registered, meaning that it may not have the correct return type. I think it does no harm to still accept it here.

What do you think? If possible I'd like to avoid SQLAlchemy version conditional code in GeoAlchemy.

Reply:
And for the record, this test is a bit of a hack, and it makes it impossible to use the form Class.geom.spatial_func to call SpatiaLite functions that don't have the ST_ prefix.

Reply:
> What do you think? If possible I'd like to avoid SQLAlchemy version conditional code in GeoAlchemy.

I would also prefer to avoid version condition. I am ok to delegate the case-sensitivity to SQLAlchemy and not fixing it in geoalchemy2. But since we know it is broken for `sqlalchemy < 1.3.4` maybe we should add a comment in the code or in the doc?

> And for the record, this test is a bit of a hack, and it makes it impossible to use the form Class.geom.spatial_func to call SpatiaLite functions that don't have the ST_ prefix.

Indeed. Maybe we should have a kind of `functions._registry` which could be a `set()` containing all the function names? And the test would just become `if name.lower() not in functions._registry`?

Reply:
> Indeed. Maybe we should have a kind of `functions._registry` which could be a `set()` containing all the function names? And the test would just become `if name.lower() not in functions._registry`?

Not sure. Because I still want to be able to do Class.geom.ST_Func although I don't have a GenericFunction class for ST_Func. We actually don't need GenericFunction classes for functions that don't return geometries.

Reply:
My suggestion is to use `if not name.lower().startswith('st_')` for all SQLAlchemy versions. What problems do you anticipate with that?

Reply:
> We actually don't need GenericFunction classes for functions that don't return geometries.

Isn't it contradictory to this: https://github.com/geoalchemy/geoalchemy2/pull/224#issuecomment-487983471? ;-P

> Not sure. Because I still want to be able to do Class.geom.ST_Func although I don't have a GenericFunction class for ST_Func.

Ok I see. So let's go for `if not name.lower().startswith('st_')`.
But in this case, what is the point to enumerate the functions that do not return a `Geometry` in `functions.py`?

> My suggestion is to use if not name.lower().startswith('st_') for all SQLAlchemy versions. What problems do you anticipate with that?

I don't anticipate anything more that what we already know, I just feel embarrassed to let a known undesired behavior (returning the wrong type for `sqlalchemy < 1.3.4` when the letter case is not exactly as expected by `geoalchemy2`).

Reply:
> Isn't it contradictory to this: #224 (comment)? ;-P

A little maybe. Currently we add GenericFunction classes for both geometry-returning and non geometry-returning functions. Adding GenericFunction classes for non geometry-returning functions is not that useful, I guess we do it just for documentation purposes.

> But in this case, what is the point to enumerate the functions that do not return a Geometry in functions.py?

Documentation I guess. But this is not a good argument. I think we can revisit that choice.

> I just feel embarrassed to let a known undesired behavior (returning the wrong type for sqlalchemy < 1.3.4 when the letter case is not exactly as expected by geoalchemy2).

But you're not fixing this undesired behavior with this SQLAlchemy version dependent code. Not for every case I mean. There is always the case where the user does func.st_buffer(geom) in her code, as opposed to Class.geom.st_buffer. That case is not detected.

So for SQLAlchemy < 1.3.4 I guess we need to live with the limitation.

Reply:
> Documentation I guess. But this is not a good argument. I think we can revisit that choice.

Ok. We will see that later then.

> But you're not fixing this undesired behavior with this SQLAlchemy version dependent code. Not for every case I mean. There is always the case where the user does func.st_buffer(geom) in her code, as opposed to Class.geom.st_buffer. That case is not detected.
So for SQLAlchemy < 1.3.4 I guess we need to live with the limitation.

You are right, I can live with that :)